### DISCUSSION TITLE
A way to retrieve a specific value without ctor overhead

### DISCUSSION THREAD

**Discussion by peppy** [realm/realm-dotnet#2773] [Q&A] [answered]
In realm-dotnet, there is a GC/alloc overhead associated with traversal. An example usage which can be quite expensive is as such:

```csharp
var cats = realm.All<Cat>();
var list = new List<string>();

foreach (var c in cats)
    list.Add(c.NameTag.Name);
```

Even though we are only interested in the cat's name, this will construct every `Cat` and every `NameTag`, incurring an alloc/GC overhead, but also a realm handle overhead.

Is there any way to do this kind of query more efficiently? I've tried to find a method using the dynamic API but it doesn't seem readily possible.

**Comment by fealebenpae** [comment, marked as answer]
There isn't a way to do this more efficiently, I'm afraid. I have had the idea of allowing realm models to be value types, which should help reduce GC overhead, but there would still be the matter of performing the P/Invoke calls to get the `Cat` and `NameTag` handles.

I think in this case supporting something like `realm.All<Cat>().Select(c => c.NameTag.Name).ToList()` would be ideal, and it'd be worth looking into what the Swift SDK did for [the Projections feature they released recently](https://github.com/realm/realm-swift/releases/tag/v10.21.0).

**Reply by peppy** [reply]
The idea of an `IQueryable` projection would be precisely what I'd be after. Fingers crossed we'll see that in some future.

This has come up a few times when trying to get our project to meet our existing performance goals (previously SQLite), but the specific case that made me post this was retrieving all `Guid` for a subscribed set in order to bring an external list into a consistent state (https://github.com/ppy/osu/pull/16523/files#diff-0b1dcf0afe266751400b0c7b7256a23274dc0cab9fc8aab278cf2ce804be079cR236-R238).

Thanks for the answer either way.

**Reply by nirinchev** [reply]
Just out of curiosity, have you measured the overhead that these allocations add to your game? I tried to reconstruct a very primitive test based on my understanding of the code in the linked PR (particularly, the hashset construction one). It looks like this:

```csharp
    class Program
    {
        static void Main(string[] args)
        {
            // Prepopulated with 24k objects
            var realm = Realm.GetInstance("beatmaps");

            var sw = new Stopwatch();
            sw.Start();

            Console.WriteLine("Starting!");

            var set = new HashSet<Guid>();
            foreach (var bm in realm.All<BeatmapInfo>())
            {
                set.Add(bm.Id);
            }

            // We don't really need gen 2 collection as the RealmObject/Handle instances should
            // be gen 0, but doesn't hurt.
            GC.Collect(2, GCCollectionMode.Forced, blocking: true);

            Console.WriteLine($"Time: {sw.ElapsedMilliseconds} ms");
        }
    }

    class BeatmapInfo : RealmObject
    {
        [PrimaryKey]
        public Guid Id { get; set; }
    }
```

This runs in ~50ms on my Intel MBP. Obviously not great and can be further optimized, but I would imagine it's fine for most people, assuming that this only happens a handful of times during their game session.

To be clear, I am a big fan of Yavor's proposal and would like to see it supported by the SDK, but also want to understand if that's a "nice to have, but can live without it" kind of optimization or it's somehow responsible for the 8 seconds load time in the video and my test is not representing the issue correctly.

**Reply by peppy** [reply]
In our case, the model that is being constructed contains a lot more than just the `Id`. In addition, right now because we want to enforce non-null properties, we also initialise some properties with placeholder, that (unfortunately) live in the parameterless constructor right now:

https://github.com/ppy/osu/blob/3e5d29ed007277014eec19a65e4958508d7d8bb3/osu.Game/Beatmaps/BeatmapInfo.cs#L49-L60

The overhead is quite large as a result (around 60-70% of time spent in GC). But I will say that it is not responsible for the 8 second load, which is us detaching entities from realm. The blocking pause at the end of the video is where this overhead is being incurred. As an aside, there were two videos there labelled "before" and "after", but to clarify since it wasn't obvious, before was using EF and after is realm, which shows them roughly tied at the moment. There's potential room for performance gains by using native realm queries rather than fully detaching, but requires a large amount of refactoring at our end due to how the display is animated and filtered.

Let me also say that this was just the most recent case I came across, but I have ran into some other cases (of data retrievals on nested `EmbeddedObjects` or instance) which exemplify the issue.

I have another more serious (crippling) realm-related performance issue to report today, but if you'd like I can provide some benchmarks and/or isolated tests for this as well.

---

Touching on the parameterless (realm) constructor again, currently we use it in a lot of places in our code, but I am actively refactoring to avoid that. Optimally we would like to be able to make the realm ctor `protected` or `private` so it can't be consumed by us, not sure if that's feasible from your end. My current workaround is to ban its use (WIP at https://github.com/ppy/osu/commit/f0f7bb97fe90960a8594bf74828c967b86015667 should better explain what i mean, and show where the overhead comes from) and allow realm to bypass the pointless initialisations as a result.

**Reply by peppy** [reply]
Here's one better example with profiling (as fixed in https://github.com/ppy/osu/pull/16522/files):


```csharp
// all code here simplified greatly, just intending to show what we're doing at a high level

public class BeatmapSetInfo : RealmObject
{
        public IList<BeatmapInfo> Beatmaps { get; } = null!;
}

public class BeatmapInfo : RealmObject
{
        public bool Hidden { get; set; }
}

...
foreach (var s in realm.All<BeatmapSetInfo>())
{
    if (beatmapSet.Beatmaps.All(b => b.Hidden))
        return null;

    ...
}
```

Preface:
- I tried using `Filter` for the hidden check and it was slower due to the tight loop (would create an execution plan per iteration). Results for that can be seen [here](https://user-images.githubusercontent.com/191335/150291038-d9b7bca5-5f94-44a4-b151-41312016a9a8.png)
- For external reasons we can't put the hidden query in the main realm enumerable (via `Filter` currently)
- There's hidden overhead as mentioned in the last post which makes this worse than the above slimmed models.

Querying live via realm:

![20220120 161137 (Parallels Desktop)](https://user-images.githubusercontent.com/191335/150290665-66582ec0-c024-4767-8163-3ab96098513b.png)

Querying post-detach (note lack of overhead from iteration):

![20220120 161328 (Parallels Desktop)](https://user-images.githubusercontent.com/191335/150290860-194d690f-2c90-4777-b418-e8cc22e51342.png)

Interestingly, a lot of the overhead is not from our parameterless constructor but the realm fetch/enumeration call stack. I think these would both be resolved with the proposed projection lookup method.

**Reply by nirinchev** [reply]
You can totally make that ctor private/protected - Realm doesn't care.

I can see how adding extra allocations in there might exacerbate the GC issue. One thing you can try is force GC at some intervals during the loop to reduce the pause at the end. That combined with reducing the extra allocations in the ctor might be enough to avoid the frame rate drop, though it won't help with the detaching issue.

And probably not worth bothering with benchmarks for this issue - I just wanted to make sure my understanding is not totally off mark.

Edit: just saw your second post, will reply to that shortly!

**Reply by peppy** [reply]
> You can totally make that ctor private/protected - Realm doesn't care.

Oh that's really cool to hear. I guess I was hung up on the [deprecated documentation](https://docs.mongodb.com/realm-legacy/docs/dotnet/latest/) which mentions this explicitly:

![20220120 163515 (Safari)](https://user-images.githubusercontent.com/191335/150293854-8e3cff2f-0846-4c2b-a399-d4e932a644ab.png)

**Reply by nirinchev** [reply]
Yeah, that's outdated - we changed the way we instantiate objects to not use reflection anymore, so we don't really need a public ctor.

Regarding your second post, do you care about returning `null` in case there are no visible beatmaps? If you don't, then you can probably use a string query to do the filtering for you. Like:

```csharp
var nonHiddenSets = realm.All<BeatmapSetInfo>().Filter("ANY Beatmaps.Hidden == false");
foreach (var s in realm.All<BeatmapSetInfo>())
{
    ...
}
```

We now have a more complete docs for the string-based query language here: https://docs.mongodb.com/realm/reference/realm-query-language, which may be helpful.

**Reply by peppy** [reply]
I think we can actually incorporate that immediately for the said case, great to know!

**Reply by peppy** [reply]
Interestingly, adding the extra query to the realm filter has quite a large performance overhead:

```csharp
for (int i = 0; i < 5; i++)
{
    var results = realmFactory.Context
                              .All<BeatmapSetInfo>()
                              .Filter("DeletePending == false && Protected == false && ALL Beatmaps.Hidden == false")
                              .ToList();
    Console.WriteLine($"result 1: {timer.ElapsedMilliseconds}");

    timer.Restart();
    results = realmFactory.Context
                          .All<BeatmapSetInfo>()
                          .Filter("DeletePending == false && Protected == false")
                          .ToList();
    Console.WriteLine($"result 2: {timer.ElapsedMilliseconds}");

    timer.Restart();
    results = realmFactory.Context
                          .All<BeatmapSetInfo>()
                          .Where(s => !s.DeletePending && !s.Protected)
                          .ToList();
    Console.WriteLine($"result 3: {timer.ElapsedMilliseconds}");
}
```

```csharp
result 1: 112
result 2: 51
result 3: 26
result 1: 129
result 2: 34
result 3: 27
result 1: 130
result 2: 30
result 3: 43
result 1: 124
result 2: 40
result 3: 45
result 1: 136
result 2: 26
result 3: 45
```

Because we are already detaching and the case of "all beatmaps being hiddden" is *very* rare, it would make more sense to do the check post-detach in this specific case, although one could argue that since we are running the realm query asynchronously if done via the `Filter` it's much-of-a-muchness.

**Reply by nirinchev** [reply]
I think you may have an error in the query there, which is likely impacting performance. In the first query, you're applying a filter of `ALL Beatmaps.Hidden == false`, which means you only want BeatmapSetInfos where all beatmaps are visible (i.e. none are hidden). This means that the query engine needs to evaluate every object in the list. If the lists are large, this will easily dominate the evaluation time. I believe (and your original code seems to support it), you actually want to get all sets that have at least one non-hidden beatmap and if the majority of your sets meet that condition and the unhidden beatmap is one of the first elements in the list, the query should complete much faster.

To be clear, collection queries are not free, so I still expect some overhead compared to querying on a bunch of properties, but they should still be faster than doing the filtering in managed code due to the elimination of managed->native transitions and temporary object allocations.

**Reply by peppy** [reply]
Good catch, double-negatives got the best of me.

Using `.Filter("DeletePending == false && Protected == false && ANY Beatmaps.Hidden == false")` now as proposed, doesn't seem to give different results though:

```csharp
result 1: 129
result 2: 47
result 3: 43
result 1: 133
result 2: 35
result 3: 50
result 1: 140
result 2: 28
result 3: 50
result 1: 143
result 2: 53
result 3: 33
result 1: 123
result 2: 50
result 3: 42
```

**Reply by nirinchev** [reply]
Surprising - I'm guessing creating the iterator on the core side was more expensive than actually iterating the objects when querying. I'll try to create a simple benchmark reproducing this and bring it up with the team to see if there are obvious areas for improvement.
